### PR TITLE
feat: GitHub ↔ Beads sync — setup, reverse sync, docs (PR 2/2)

### PR DESCRIPTION
**Pull request by harshanandak**
## Summary

PR 2/2 for the GitHub ↔ Beads bidirectional issue sync epic (`forge-d2cl`). Builds on the core sync engine from PR #71.

- **T10: Forge setup integration** — `scaffoldGithubBeadsSync()` in `lib/setup.js`, interactive prompt in `bin/forge.js` during `configureExternalServices`. Copies workflow YAML, config JSON, and mapping template. Never overwrites existing files. Skipped in `--quick`/`--skip-external` modes.
- **T11: Beads→GitHub reverse sync** — Phase 2 of the sync: when `.beads/` changes on push to master, detects issues that transitioned to `closed` and closes the linked GitHub issue via `gh api`. Loop guard prevents infinite cycles. New workflow: `.github/workflows/beads-to-github.yml`.
- **T12: Documentation** — `docs/BEADS_GITHUB_SYNC.md` (architecture, setup, config reference, security, troubleshooting, fork behavior), PR template update, AGENTS.md pointer.
- **Pre-existing test fix** — Increased timeout for `dep-guard.sh check-ripple` tests that flaked on Windows CI.

## Design Doc
See: `docs/plans/2026-03-21-github-beads-sync-design.md`

## Beads Issue
Related: `forge-d2cl` (GitHub ↔ Beads bidirectional issue sync)

## Key Decisions
1. **Create-only sync (Option D)** — No field updates after creation, avoids conflict resolution entirely
2. **Belt-and-suspenders mapping (Option C)** — Bot comment (human-visible) + mapping file (fast lookup, debuggable)
3. **Configurable label mapping** — `sync-config.json` with user-customizable label→type/priority maps
4. **Fail-closed config validation** — Invalid `publicRepoGate` values throw errors (not silently default)
5. **Convention over configuration** — Works with zero setup via GitHub's native auto-add-to-project

## TDD Test Coverage
- Unit tests: 86 tests across 7 modules (config, mapping, comment, github-api, sanitize, run-bd, label-mapper)
- Integration tests: 7 tests with fixture payloads (opened, closed, malicious, auth gate)
- Orchestration tests: 12 tests for handleOpened/handleClosed
- Workflow validation: 27 tests for YAML security constraints
- Reverse sync tests: 20 tests for Phase 2 detection/closing
- Setup tests: 7 tests for scaffold function
- **Total: 159 new tests, all passing ✓**

## Security Review
- OWASP Top 10: Full analysis in `docs/plans/2026-03-21-issue-sync-owasp-analysis.md`
- A03 Injection: `execFileSync` with array args, `sanitizeTitle`/`sanitizeBody` strip shell metacharacters
- A05 Misconfiguration: Fail-closed `validateConfig()` with "did you mean?" suggestions
- A08 Data Integrity: `concurrency` group prevents mapping file corruption
- SHA-pinned third-party Actions (checkout, setup-bun)
- No `${{ }}` interpolation of untrusted data in `run:` blocks

## Test Plan
- [x] Lint passing (0 errors, 0 warnings)
- [x] Tests passing (1887 pass, 0 fail)
- [x] Security audit clean
- [x] OWASP review completed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details><summary><h3>Confidence Score: 5/5</h3></summary>

- Safe to merge — prior scaffold gap is resolved, logic is sound, and the only remaining item is a minor path-trigger scoping suggestion.
- The primary blocker from the previous review (missing Phase 2 workflow + script modules in scaffold) is addressed. The reverse-sync implementation handles all significant edge cases (empty old content, initial push, per-issue error isolation, loop guard). Test coverage is comprehensive (20 unit tests for reverse sync, 7 for scaffold). The single P2 comment (`.beads/**` vs `.beads/issues.jsonl`) is a non-blocking efficiency improvement that doesn't affect correctness.
- No files require special attention.
</details>

<details open><summary><h3>Comments Outside Diff (3)</h3></summary>

1. `bin/forge.js`, line 110-117 ([link](https://github.com/harshanandak/forge/blob/1bca335149b7859d332cc3f1fe9ad76f9fe8d59f/bin/forge.js#L110-L117)) 

   <a href="#">P2</a> **Missing error handling around scaffold**

   `scaffoldGithubBeadsSync` calls `fs.promises.copyFile(file.src, destPath)` for each template file. If any source file is absent (e.g. a partial npm install), the function throws `ENOENT` and propagates uncaught through `configureExternalServices`, crashing the setup flow with no user-friendly message. Wrapping in a try-catch lets you surface a clear error instead:

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: bin/forge.js
   Line: 110-117

   Comment:
   **Missing error handling around scaffold**

   `scaffoldGithubBeadsSync` calls `fs.promises.copyFile(file.src, destPath)` for each template file. If any source file is absent (e.g. a partial npm install), the function throws `ENOENT` and propagates uncaught through `configureExternalServices`, crashing the setup flow with no user-friendly message. Wrapping in a try-catch lets you surface a clear error instead:

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. `test/scripts/github-beads-sync/reverse-sync.test.js`, line 1079-1088 ([link](https://github.com/harshanandak/forge/blob/1bca335149b7859d332cc3f1fe9ad76f9fe8d59f/test/scripts/github-beads-sync/reverse-sync.test.js#L1079-L1088)) 

   <a href="#">P2</a> **Loop guard tests don't exercise the actual shell pattern**

   The tests in this block verify a JavaScript `String.prototype.startsWith` call, not the bash glob pattern `[[ "$COMMIT_MSG" == chore\(beads\):* ]]` from the workflow. The comment "This tests the logic the workflow YAML uses" is misleading — if the bash pattern in `.github/workflows/beads-to-github.yml` were changed or broken, these tests would still pass.

   Consider either:
   - Removing these trivial JS tests (they add no real coverage), or
   - Adding a shellcheck/workflow-validation test against the actual YAML file, similar to the existing workflow validation tests mentioned in the PR description, to make it harder for the shell pattern to silently diverge.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: test/scripts/github-beads-sync/reverse-sync.test.js
   Line: 1079-1088

   Comment:
   **Loop guard tests don't exercise the actual shell pattern**

   The tests in this block verify a JavaScript `String.prototype.startsWith` call, not the bash glob pattern `[[ "$COMMIT_MSG" == chore\(beads\):* ]]` from the workflow. The comment "This tests the logic the workflow YAML uses" is misleading — if the bash pattern in `.github/workflows/beads-to-github.yml` were changed or broken, these tests would still pass.

   Consider either:
   - Removing these trivial JS tests (they add no real coverage), or
   - Adding a shellcheck/workflow-validation test against the actual YAML file, similar to the existing workflow validation tests mentioned in the PR description, to make it harder for the shell pattern to silently diverge.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

3. `.github/workflows/beads-to-github.yml`, line 4-7 ([link](https://github.com/harshanandak/forge/blob/3d3bff45b1ee691465b71ba97071b4b5ac67a8fd/.github/workflows/beads-to-github.yml#L4-L7)) 

   **Path trigger broader than necessary**

   The `paths` filter catches all changes to `.beads/**`, but the reverse-sync logic only ever reads `.beads/issues.jsonl`. Any other file written under `.beads/` (e.g., settings, metadata) will kick off a full checkout + Node.js run that immediately produces `{"closed":[],"skipped":[],"errors":[]}` and exits 0 — wasting runner minutes for every such change.

   Narrowing the trigger means the workflow only fires when the file it actually processes changes:

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: .github/workflows/beads-to-github.yml
Line: 4-7

Comment:
**Path trigger broader than necessary**

The `paths` filter catches all changes to `.beads/**`, but the reverse-sync logic only ever reads `.beads/issues.jsonl`. Any other file written under `.beads/` (e.g., settings, metadata) will kick off a full checkout + Node.js run that immediately produces `{"closed":[],"skipped":[],"errors":[]}` and exits 0 — wasting runner minutes for every such change.

Narrowing the trigger means the workflow only fires when the file it actually processes changes:

```suggestion
    paths:
      - '.beads/issues.jsonl'
```

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: ["fix: add scaffold er..."](https://github.com/harshanandak/forge/commit/3d3bff45b1ee691465b71ba97071b4b5ac67a8fd)</sub>

### CODE REVIEW

File: lib/setup.js [hunk 1]
```diff
@@ -109,10 +109,66 @@ async function markStepComplete(projectPath, stepName) {
...
+    {
+      src: path.join(pkgDir, 'scripts', 'github-beads-sync.config.json'),
+      dest: path.join('scripts', 'github-beads-sync.config.json'),
+    },
+    {
+      src: null, // generated in-place (empty mapping template)
+      dest: path.join('.github', 'beads-mapping.json'),
+      content: '{}',
+    },
+  ];
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/setup.js:156]: ✅ Fixed: Scaffold now includes Phase 2 workflow (beads-to-github.yml) + all 10 sync script modules. Total: 14 files scaffolded.

Changed: lib/setup.js — added beads-to-github.yml and syncScripts array to files list
Reason: Users enabling sync via forge setup need all files, not just Phase 1
Commit: 1bca335

File: scripts/github-beads-sync/reverse-sync-cli.mjs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+const result = handleBeadsClosed(oldContent, newContent);
+console.log(JSON.stringify(result, null, 2));
+
+if (result.errors.length > 0) {
+  console.error(`${result.errors.length} issue(s) failed to close on GitHub`);
+  process.exit(1);
+}
+
+console.log(`Closed ${result.closed.length} GitHub issue(s), skipped ${result.skipped.length}`);
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply scripts/github-beads-sync/reverse-sync-cli.mjs:31]: ✅ Fixed: Added test/scripts/github-beads-sync/reverse-sync-cli.test.js with 5 tests covering missing args (exit 1), empty files (exit 0), closure detection, and skip-no-URL path.

Changed: New test file for CLI entry point
Reason: Repository compliance requires test file per source file
Commit: 1bca335

File: .github/workflows/beads-to-github.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+          fetch-depth: 2
+
+      - name: Detect and close
+        if: env.SKIP != 'true'
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          # Get old and new versions of .beads/issues.jsonl
+          OLD_CONTENT=$(git show HEAD~1:.beads/issues.jsonl 2>/dev/null || echo "")
+          NEW_CONTENT=$(cat .beads/issues.jsonl 2>/dev/null || echo "")
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply .github/workflows/beads-to-github.yml]: ✅ Fixed: Workflow now uses github.event.before (pre-push SHA) instead of HEAD~1. Added fallback for initial push (all-zero SHA). Changed fetch-depth to 0 to ensure the before SHA is available.

Changed: .github/workflows/beads-to-github.yml — BEFORE_SHA env var from github.event.before
Reason: HEAD~1 only sees the last commit; multi-commit pushes miss earlier transitions
Commit: 1bca335

File: lib/setup.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (harshanandak) [reply lib/setup.js:156]: ✅ Fixed: Same fix as Greptile P1 above — scaffold now includes all 14 files (2 workflows + config + mapping + 10 script modules). Tests updated to verify.

Changed: lib/setup.js, test/setup-github-sync.test.js
Commit: 1bca335
